> 原文链接: https://matklad.github.io/2022/03/26/self-modifying-code.html
>
> **翻译：[朕与将军解战袍](https://github.com/a1393323447)**
>
> 选题：[朕与将军解战袍](https://github.com/a1393323447)
>
> 本文由 [Rustt](https://Rustt.org) 翻译，[StudyRust](https://studyrust.org) 荣誉推出

# 自修改代码

这篇文章和 JIT 之类、在运行时为机器码打补丁的技术无关（尽管这些技术很炫酷）。相反，这篇文章会介绍一种 可爱/可怕 的 trick/hack 。如果你不是宏的超级粉丝，你可以用它来生成**源**代码。最终的技术将独立于任何特定的编程语言，但先导技术（the lead-up）会是 Rust 特有的。这种模式可以被应用到各种各样的任务中，但我们将通过一个典例来研究各种不同的解决方案。

## 问题

有一个没有字段的枚举，用于表示各种错误条件：

```rust
#[derive(Debug, Clone, Copy)]
pub enum Error {
  InvalidSignature,
  AccountNotFound,
  InsufficientBalance,
}
```

我预计这个枚举经常会被更改，会变大很多。尽管最初版本就已经有 6 个枚举成员了！简单起见，我只展示了一个子集。

为了序列化，我希望 `error` 能在枚举和错误码之间互相转换。其中一个转换方向很容易实现，Rust 有满足这个要求的内置机制：

```rust
impl Error {
  pub fn as_code(self) -> u32 {
    self as u32
  }
}
```

另一个转换方向就恼人的多了：Rust 还不能自动处理这个问题（尽管现在有一个正在推进的 PR 增加了这一点！），所以我们不得不自己写一些代码：

```rust
impl Error {
  pub fn as_code(self) -> u32 {
    self as u32
  }

  pub fn from_code(code: u32) -> Option<Error> {
    let res = match code {
      0 => Error::InvalidSignature,
      1 => Error::AccountNotFound,
      2 => Error::InsufficientBalance,
      _ => return None,
    };
    Some(res)
  }
}
```

现在，考虑到这个类型会被频繁更改，上面的做法就是自找麻烦！`match` 语句和枚举定义很容易不同步!

我们该做什么?我们**能**做什么?

## 极简方案

现在，经验丰富的 Rust 开发者可能已经开始考虑宏（甚至是一些具体的和宏有关的 crate）了。然后我们成功解决问题了！但请先看一下，当不想写宏的时候（我一般都是这样），我通常是怎么解决这个问题的。

整体思路是：让编译器告诉我们枚举中的元素个数，让我们能够做一些健全性检查。可以通过在枚举的末尾添加一个假元素实现：

```rust
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Error {
  InvalidSignature,
  AccountNotFound,
  InsufficientBalance,
  __LAST,
}

impl Error {
  const ALL: [Error; Error::__LAST as usize] = [
    Error::InvalidSignature,
    Error::AccountNotFound,
    Error::InsufficientBalance,
  ];

  pub fn from_code(code: u32) -> Option<Error> {
    Error::ALL.get(code as usize).copied()
  }
  pub fn as_code(self) -> u32 {
    Error::ALL
      .into_iter()
      .position(|it| it == self)
      .unwrap_or_default() as u32
  }
}
```

现在，如果我们增加了一个新的错误枚举成员，却忘了更新 `ALL` 数组，那么这段代码不能通过编译——这正是我们想要的提醒。这种方法的主要缺点是：`__LAST` 必须要存在。这对于内部的东西来说不错，但是对于一个公共的、干净的 API 来说就不太好了。

## 极简的宏

现在，我们来看一下宏，从我能想到的最简单的开始！

```rust
define_error![
  InvalidSignature,
  AccountNotFound,
  InsufficientBalance,
];

impl Error {
  pub fn as_code(self) -> u32 {
    self as u32
  }
}
```

很简单吧？那我们来看一遍 `define_error!` 的定义：

```rust
macro_rules! define_error {
  ($($err:ident,)*) => {
    #[derive(Debug, Clone, Copy, PartialEq, Eq)]
    pub enum Error {
      $($err,)*
    }

    impl Error {
      pub fn from_code(code: u32) -> Option<Error> {
        #![allow(non_upper_case_globals)]
        $(const $err: u32 = Error::$err as u32;)*
        match code {
          $($err => Some(Error::$err),)*
          _ => None,
        }
      }
    }
  };
}
```

这真的是...一堆乱码！声明宏机制相对来说缺乏表达力，因此需要发挥创造性才能得到你想要的效果。理想情况下我会写：

```rust
match code {
  0 => Error::InvalidSignature,
  1 => Error::AccountNotFound,
  2 => Error::InsufficientBalance,
}
```

唉，在宏中计数是可行的，却是很不简单的，算是一个子谜题（subpuzle）！比起解谜，我用了下面的方法：

```rust
const InvalidSignature: u32 = Error::InvalidSignature as u32;
match {
  InvalidSignature => Error::InvalidSignature,
}
```

然后我不得不使用 `#![allow(non_upper_case_globals)]` ，以消除编译器警告。

## 地道的宏

宏的严重问题是，它不仅是内部实现，还是 baroque[^1] 的！调用点也很神秘！假设我们刚接触代码库，遇到以下代码片段：

```rust
define_error![
  InvalidSignature,
  AccountNotFound,
  InsufficientBalance,
];

impl Error {
  pub fn as_code(self) -> u32 {
    self as u32
  }
}
```

我会思考一个问题 ”这个 `Error` 是个什么东西？“ 。 还好，我们生活在强大的 IDE 时代，所以我们可以通过 “转向定义（goto definition）” 找到答案，对吧?

<img src="https://raw.githubusercontent.com/rustt-org/rustt-assets/main/20220408 Self Modifying Code/gotodef-macro-1.gif" />

好吧，和我们想的不一样。IDE 表示 `Error` 是由调用的宏里面的**某些东西**生成的。答案正确，但没什么用！所以我还是不得不去看 `define_error` 宏的定义，然后为了知道外部公共 API （例如，`Error` 是一个公共的枚举）怎么用还要搞清楚这个宏是怎么工作的。在这里，声明宏的谜题性质更加突出了。要弄清楚怎么样用宏这种受限制的语言来表达你思想已经够困难了。要搞清楚宏作者的意图更是加倍的困难。毕竟你不能直接看到他们脑子里想法，只能看到宏的实现。

补救办法是让宏的输入看起来像想要生成的代码。像下面这样：

```rust
define_error![
  #[derive(Debug, Clone, Copy, PartialEq, Eq)]
  pub enum Error {
    InvalidSignature,
    AccountNotFound,
    InsufficientBalance,
  }
];

impl Error {
  pub fn as_code(self) -> u32 {
    self as u32
  }
}
```

这种方法确实对 IDE 和人来说都稍微友好一些：

<img src="https://raw.githubusercontent.com/rustt-org/rustt-assets/main/20220408 Self Modifying Code/gotodef-macro-2.gif" />

代价是一个更加复杂的宏实现。一般来说，宏要做两件事：解析任意的 `token` 输入流，然后输出有效的 Rust 代码。解析通常是更复杂的任务。这就是为什么在极简实现中，我们只使用了最简单的语法，仅仅是一个标识符列表。然而，如果想让宏的输入看起来更像 Rust ，我们必须解析 Rust 的一个子集。这更复杂：

```rust
macro_rules! define_error {
  (
    $(#[$meta:meta])*
    $vis:vis enum $Error:ident {
      $($err:ident,)*
    }
  ) => {
    $(#[$meta])*
    $vis enum $Error {
      $($err,)*
    }

    impl Error {
      pub fn from_code(code: u32) -> Option<Error> {
        #![allow(non_upper_case_globals)]
        $(const $err: u32 = $Error::$err as u32;)*
        match code {
          $($err => Some($Error::$err),)*
          _ => None,
        }
      }
    }
  };
}

define_error![
  #[derive(Debug, Clone, Copy, PartialEq, Eq)]
  pub enum Error {
    InvalidSignature,
    AccountNotFound,
    InsufficientBalance,
  }
];
```

我们必须小心翼翼地处理所有的可见性声明和属性。即使我们这样做了，这些类似 Rust 的输入语法也只是在表面上和输出的 Rust 代码有联系。这基本上是一种假象，和在这里使用 Haskell 语法没有太大区别：

```rust
macro_rules! define_error {
  (
    data $Error:ident = $err0:ident $(| $err:ident)*
      $(deriving ($($derive:ident),*))?
  ) => {
    $(#[derive($($derive),*)])?
    enum $Error {
      $err0,
      $($err,)*
    }

    impl Error {
      pub fn from_code(code: u32) -> Option<Error> {
        #![allow(non_upper_case_globals)]
        const $err0: u32 = $Error::$err0 as u32;
        $(const $err: u32 = $Error::$err as u32;)*
        match code {
          $err0 => Some($Error::$err0),
          $($err => Some($Error::$err),)*
          _ => None,
        }
      }
    }
  };
}

define_error![
  data Error = InvalidSignature | AccountNotFound | InsufficientBalance
    deriving (Debug, Clone, Copy, PartialEq, Eq)

];

impl Error {
  pub fn as_code(self) -> u32 {
    self as u32
  }
}
```

## 属性宏

通过切换到派生宏，我们可以增加宏输入和宏输出之间的保真度。与类似函数的宏相比，派生宏要求它们的输入在语法上甚至语义上都是有效的。

结果看起来像这样：

```rust
use macros::FromCode;

#[derive(FromCode, Debug, Clone, Copy, PartialEq, Eq)]
enum Error {
  InvalidSignature,
  AccountNotFound,
  InsufficientBalance,
}

impl Error {
  pub fn as_code(self) -> u32 {
    self as u32
  }
}
```

而在这里 `Error` 是一个直白的、简单的枚举！不再是一个披着 `enum` 外皮的外星怪兽了。

而宏的实现看起来也很不错。再次感谢 @dtolnay 的优雅的 API 设计：

```rust
use proc_macro::TokenStream;
use quote::quote;
use syn::{parse_macro_input, DeriveInput};

#[proc_macro_derive(FromCode)]
pub fn from_code(input: TokenStream) -> TokenStream {
  let input = parse_macro_input!(input as DeriveInput);
  let error_name = input.ident;
  let enum_ = match input.data {
    syn::Data::Enum(it) => it,
    _ => panic!("expected an enum"),
  };

  let arms =
    enum_.variants.iter().enumerate().map(|(i, var)| {
      let i = i as u32;
      let var_name = &var.ident;
      quote! {
        #i => Some(#error_name::#var_name),
      }
    });

  quote! {
    impl #error_name {
      pub fn from_code(code: u32) -> Option<#error_name> {
        match code {
          #(#arms)*
          _ => None,
        }
      }
    }
  }
  .into()
}
```

和声明性宏不同，我们只是直接表达想要产出的语法——一个对连续自然数的 `match` 语句 。

这种方法最大的缺点是：在调用处，我们不知道宏生成的额外 API 。如果使用声明宏，你可以注意到在同一个文件中有一个 `pub fn from_code` ，然后猜测它是 API 的一部分，而过程宏则是在一个完全不同的 crate 中！虽然过程宏可以极大地改善使用和实现宏的人机工程学（尽管增加了编译时间），但对于读者来说，它们可以说比声明式宏更不透明。

[^1]: 译者注：baroque 是一种艺术风格。形容词有“俗丽凌乱”的含义。

## 自修改代码

最后，我们来看一下一开始说的 hacky solution :-)  虽然，你可能已经注意到，我不是宏的超级粉丝，我喜欢普通的、老式的代码生成——文本输入，文本输出。与高级宏系统相比，文本操作可以说是更 worse-is-better 。

所以我们要做的是：

- 以字符串的形式读取带有枚举定义的文件（ `file!` 宏在这一步很有用）
- 用简单的字符串分割（ 函数 `str::split_once` 也叫 cut ，会作为解析器）来“解析”枚举定义
- 通过连接字符串生成我们想要的代码
- 将生成的代码粘贴到一个特殊标记的位置
- 如果有更改，就重写文件
- 我们会通过 `#[test]` 触发上述过程！

```rust
#[derive(Debug, Clone, Copy)]
pub enum Error {
  InsufficientBalance,
  InvalidSignature,
  AccountNotFound,
}

impl Error {
  fn as_code(self) -> u32 {
    self as u32
  }

  fn from_code(code: u32) -> Option<Error> {
    let res = match code {
      // region:sourcegen
      0 => Error::InsufficientBalance,
      1 => Error::InvalidSignature,
      2 => Error::AccountNotFound,
      // endregion:sourcegen
      _ => return None,
    };
    Some(res)
  }
}

#[test]
fn sourcegen_from_code() {
  let original_text = std::fs::read_to_string(file!()).unwrap();
  let (_, variants, _) =
    split_twice(&original_text, "pub enum Error {\n", "}")
      .unwrap();

  let arms = variants
    .lines()
    .map(|line| line.trim().trim_end_matches(','))
    .enumerate()
    .map(|(i, var)| format!("      {i} => Error::{var},\n"))
    .collect::<String>();

  let new_text = {
    let start_marker = "      // region:sourcegen\n";
    let end_marker = "      // endregion:sourcegen\n";
    let (prefix, _, suffix) =
      split_twice(&original_text, start_marker, end_marker)
        .unwrap();
    format!("{prefix}{start_marker}{arms}{end_marker}{suffix}")
  };

  if new_text != original_text {
    std::fs::write(file!(), new_text).unwrap();
    panic!("source was not up-to-date")
  }
}

fn split_twice<'a>(
  text: &'a str,
  start_marker: &str,
  end_marker: &str,
) -> Option<(&'a str, &'a str, &'a str)> {
  let (prefix, rest) = text.split_once(start_marker)?;
  let (mid, suffix) = rest.split_once(end_marker)?;
  Some((prefix, mid, suffix))
}
```

这就是完整的模式！请注意，和其他解决方案不同的是，这种生成代码的方法的工作方式是一清二楚的，是一段你可以直接跳转到定义的代码。

这种“宏”的代码也是很好懂的——其实就是一些字符串操作。更重要的是，你可以通过运行测试轻松地看到它是怎么工作的！

“读取并更新自己的源代码”听起来有点饶！实现却是非常小的，并且只用到了标准库，所以也应该很好懂。

与宏不同，它不会在编译时强制生成的代码是新的。如果你更新了 `Error` 的定义，就需要重新运行测试，生成新的代码。而这会在测试的时候发现。请注意一个重要的细节——测试只会在发生更改的时候，尝试更新源代码。也就是说，只有在开发的时候才需要一个可写的 `src/` （目录）。

